### ISSUE TITLE
passport-facebook-token returning a 500 duplicate error on second login

### ISSUE THREAD

**Issue by ricick** [jaumard/trailpack-passport#206]
I've installed passport-facebook-token with the following strategy config:

'facebook-token': {
      protocol: 'oauth2',
      strategy: require('passport-facebook-token'),
      options: {
        clientID: cogniss.facebook.appId,
        clientSecret: cogniss.facebook.secret,
        profileFields: ['id', 'emails', 'name'],
        scope: ['email'] // email is necessary for login behavior
      }
    }

On the second call to "/api/v1/auth/facebook-token/callback?access_token=MYFBTOKEN" i get a 500 server error with the message 'E11000 duplicate key error collection: cogniss.user index: username_1 dup key: { : null }'

I'm using trailpack mongoose as my storage, but it's working fine with the other passport login methods. 

Incidentally i got stuck for ages calling "/api/v1/auth/facebook-token?access_token=MYFBTOKEN" and wondering why I got a 404 back because the controller assumes the provider will redirect the user somewhere instead of just logging them in.

**Comment by ricick** [comment]
I've tracked down the issue to a combination of both modules. I'm not sure which one is at fault.

On line 56 of lib/passports.js we call the following
const provider = profile.provider || query.provider
this returns "facebook" when the provider property on the passport in the db is "facebook-token" so app.services.FootprintService.find on line 84 will never return an existing passport. query.provider here is "facebook-token" but is ignored because of the pipes.

The 500 error is then caused because passport tries to create a duplicate index when saving the "new" passport.

On line 168 of lib/index.js in passport-facebook-token when constructing the profile object you can see that the provider property is set to "facebook"

This can all be fixed by reversing the order of the pipes on line 56 of lib/passports.js to prefer the query
const provider = query.provider || profile.provider, or by changing  passport-facebook-token's profile construction to set the provider property as "facebook-token".

Which one is the correct fix. If it's a problem with them I'll raise a ticket there.

**Comment by jaumard** [comment]
Hello, in that case did you try to rename your your strategy to "facebook" instead of "facebook-token" ? My implementation is based on the fact that the name of your strategy you configured match what return this strategy. Let me know if it fix your problem.

**Comment by ricick** [comment]
Looks like that doesn't work. The entry in the passports table still has facebook-token as it's provider

**Comment by jaumard** [comment]
Then I guess the correct solution is to switch to `query.provider || profile.provider` but it need to be test with several third parties to be sure it's working for everyone or it will break other projects ... :/ 
Can you test that it's fix your problem to switch the provider ?

**Comment by ricick** [comment]
I have integration tests running with my project for login with facebook-token, jwt, and local auth. All three work with that change.

**Comment by jaumard** [comment]
Yes for jwt and local I'm not surprised it's working, I'm more concern about GitHub, Twitter, G+ ect... Jwt and local doesn't use same pipe to connect as there no third party calls

**Comment by ricick** [comment]
Ok I'll see if we're going to need g+ login for our product and get that in there and tested

**Comment by jaumard** [comment]
I'm currently testing your solution with few strategies, if there no problem I'll push a new version with the fix :)

**Comment by jaumard** [comment]
So there more than just the provider problem unfortunately, there some inconsistency between the ORM trailpacks which make the third party not working for sequelize and mongoose. I fixed it for sequelize but for mongoose I don't have enough knowledge about it to fix it.
I push everything here: 
https://github.com/jaumard/trailpack-passport/tree/feature/third_party
I think the last problem is here exactly https://github.com/jaumard/trailpack-passport/blob/feature/third_party/lib/passports.js#L140 I manage to get the user id for waterline and sequelize, but can't for mongoose so it still fail on it. Can you help with this ?

**Comment by ricick** [comment]
Yup I can take a look at this. I'm using mongoose so can easily test.

**Comment by jaumard** [comment]
I have added some unit tests for third party connexion and they didn't pass for mongoose and sequelize currently. I'll check for sequelize as I know it. I let you mongoose :) thanks a lot

**Comment by jaumard** [comment]
hey @ricick did you had the chance to take a look ?
